contracts/SimpleVaultMigrator.sol

```diff
@@ -0,0 +1,264 @@
+    function deposit(uint256 _amount, address recipient)
+        external
+        returns (uint256 shares);
+
+    function withdraw(uint256 _shares) external;
+
+    function token() external view returns (address);
+
+    function permit(
+        address owner,
+        address spender,
+        uint256 value,
+        uint256 deadline,
+        bytes calldata signature
+    ) external returns (bool);
+}
+
+interface IRegistry {
+    function latestVault(address token) external view returns (address);
+}
+
+contract SimpleVaultMigrator {
+    using SafeERC20 for IERC20;
+    using SafeERC20 for IVaultAPI;
+
+    /* ========== STATE VARIABLES ========== */
+
+    /// @notice Governance can update the registry and sweep stuck tokens.
+    address public governance;
+
+    /// @notice New address must be set by current gov and then accepted to transfer power.
+    address public pendingGovernance;
+
+    /// @notice Vault registry to pull info about yearn vaults. This will vary based on network.
+    address public registry;
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(address _governance, address _registry) {
+        require(_governance != address(0), "Governance cannot be 0");
+        require(_registry != address(0), "Registry cannot be 0");
+        governance = _governance;
+
+        registry = _registry;
+    }
+
+    /* ========== EVENTS ========== */
+    event NewGovernance(address indexed governance);
+
+    event NewRegistry(address indexed registry);
+
+    event SuccessfulMigration(
+        address indexed user,
+        address indexed vaultFrom,
+        address indexed vaultTo,
+        uint256 migratedAmount
+    );
+
+    /* ========== MODIFIERS ========== */
+    modifier onlyGovernance {
+        require(msg.sender == governance, "Sender must be governance");
+        _;
+    }
+
+    modifier onlyPendingGovernance {
+        require(
+            msg.sender == pendingGovernance,
+            "Sender must be pending governance"
+        );
+        _;
+    }
+
+    modifier checkVaults(address vaultFrom, address vaultTo) {
+        require(
+            IVaultAPI(vaultFrom).token() == IVaultAPI(vaultTo).token(),
+            "Vaults must have the same token"
+        );
+        require(
+            IRegistry(registry).latestVault(IVaultAPI(vaultFrom).token()) ==
+                vaultTo,
+            "Target vault should be the latest for token"
+        );
+        _;
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    /**
+     * @notice Migrate all of our vault shares to the newest vault version.
+     * @dev Throws if the vaults do not have the same want, or if our
+     *  target is not the newest version
+     * @param vaultFrom The old vault we are migrating from.
+     * @param vaultTo The new vault we are migrating to.
+     */
+    function migrateAll(address vaultFrom, address vaultTo) external {
+        uint256 shares = IVaultAPI(vaultFrom).balanceOf(msg.sender);
```

Comment:
Should we add a `require(shares > 0, 'shares > 0');`?

Reply:
is there any specific behavior you're worried about if a user tries to migrating 0? I supposed we could add that require statement in `_migrate` if we wanted

Reply:
from what I can tell trying to do 0 would revert because of the `safeTransferFrom` call in `_migrate` anyway, so this require statement would just add a bit more context for the revert

Reply:
Yes, I guess it wouldn't fail in the `safeTransferFrom`. As far as I see, that only verifies the returned value [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v4.7/contracts/token/ERC20/utils/SafeERC20.sol#L36). But afaik it should not be a critical issue.